src/clr-angular/data/datagrid/providers/columns.service.ts

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { Injectable, OnDestroy } from '@angular/core';
+import { BehaviorSubject, Subscription } from 'rxjs';
+import { DatagridColumnState } from '../interfaces/column-state.interface';
+import { DatagridRenderStep } from '../enums/render-step.enum';
+import { DatagridRenderOrganizer } from '../render/render-organizer';
+
+@Injectable()
+export class ColumnsService implements OnDestroy {
+  subscriptions: Subscription[] = [];
+  columns: BehaviorSubject<DatagridColumnState>[] = [];
+
+  constructor(private organizer: DatagridRenderOrganizer) {
+    this.subscriptions.push(
+      this.organizer.filterRenderSteps(DatagridRenderStep.CLEAR_WIDTHS).subscribe(() => this.reset())
+    );
+  }
+
+  ngOnDestroy() {
+    this.subscriptions.forEach(sub => sub.unsubscribe());
+  }
+
+  private reset() {
+    this.columns.forEach((column, index) => {
+      this.emitStateChange(index, { width: 0 });
+    });
+  }
+
+  // Helper method to emit a change to a column only when there is an actual diff to process for that column
+  emitStateChange(columnIndex: number, diff: Partial<DatagridColumnState>) {
+    if (!this.columns[columnIndex]) {
+      return;
+    }
+    const current = this.columns[columnIndex].value;
```

Comment:
I think we will try to access the `.value` many times from the `ColumnService`. So I think it's ideal to have a method for this like:

```ts
inspectStateOf(columnIndex: number) {
   return this.columns[columnIndex].value
}
```

Reply:
I think we can optimize this later if we do, otherwise its premature at this stage what we need here.